loki/visitors/pretty_visualize.py

```diff
@@ -0,0 +1,349 @@
+        """
+        return self._symgen
+
+    def join_items(self, items, sep=", ", separable=True):
+        """
+        Concatenate a list of items into :any:`JoinableStringList`.
+
+        The return value can be passed to :meth:`format_line` or
+        :meth:`format_node` or converted to a string with `str`, using
+        the :any:`JoinableStringList` as an argument.
+        Upon expansion, lines will be wrapped automatically to stay within
+        the linewidth limit.
+
+        Parameters
+        ----------
+        items : list
+            The list of strings to be joined.
+        sep : str, optional
+            The separator to be inserted between items.
+        separable : bool, optional
+            Allow line breaks between individual :data:`items`.
+
+        Returns
+        -------
+        :any:`JoinableStringList`
+        """
+        return JoinableStringList(
+            items,
+            sep=sep,
+            width=self.linewidth,
+            cont="\n",
+            separable=separable,
+        )
+
+    def format_node(self, name, *items):
+        """
+        Default format for a node.
+
+        Creates a string of the form ``<name[, attribute, attribute, ...]>``.
+        """
+        content = ""
+        if items:
+            content = self.format_line("<", name, " ", self.join_items(items), ">")
+        else:
+            content = self.format_line("<", name, ">")
+
+        # disregard all quotes to ensure nice graphviz label behaviour
+        return content.replace('"', "")
+
+    def format_line(self, *items, comment=None, no_wrap=False):
+        """
+        Format a line by concatenating all items and applying indentation while observing
+        the allowed line width limit.
+
+        Note that the provided comment will simply be extended to the line and no line
+        width limit will be enforced for that.
+
+        :param list items: the items to be put on that line.
+        :param str comment: an optional inline comment to be put at the end of the line.
+        :param bool no_wrap: disable line wrapping.
+
+        :return: the string of the current line, potentially including line breaks if
+                 required to observe the line width limit.
+        :rtype: str
+        """
+
+        if no_wrap:
+            # Simply concatenate items and extend the comment
+            line = "".join(str(item) for item in items)
+        else:
+            # Use join_items to concatenate items
+            line = str(self.join_items(items, sep=""))
+        if comment:
+            return line + comment
+        return line
+
+    def visit_all(self, item, *args, **kwargs):
+        """
+        Convenience function to call :meth:`visit` for all given arguments.
+
+        If only a single argument is given that is iterable,
+        :meth:`visit` is called on all of its elements instead.
+        """
+        if is_iterable(item) and not args:
+            return list(
```

Comment:
What effect does the nesting of `list(chain.from_iterable(as_tuple(...)))` have? It feels like it's building multple times the sequence but I'm probably missing something here.

Reply:
Was unnecessary 8b566dd2c72513270d3db8721b1e7129847b1d08